### PR TITLE
Fixes to remove Crease Defender Warnings in simulated tests

### LINKED ISSUE #1805: Add fixes to reduce defender crease warnings
### Description of the task

As discussed in #1712, there are a couple of things we can do to stop 
 ``WARNING [crease_defender_tactic.cpp->calculateDesiredState:114] Could not draw a line from the goalie to the ball through the defender crease``
 from showing up when we run simulated tests. Implement these changes. 

### Acceptance criteria

- [ ] in simulated tests, assign goalie to a position in the defender area 
- [ ] increase the margins for `almostequals` for when it is called from within the crease_defender_tactic

### PR DESCRIPTION
**Pull request by amrm3lm**
### Description
Currently, some simulated tests produce a warning from the `crease_defender_tactic`, this pr addresses that by doing two things:

- Changed the start position of the goalkeeper in some simulated tests
- called `almosteauals` from with values of `fixed_epsilon = 1e-9` and `ulps_epsilon = 200` , compared to the default `1e-15` and `10`. 

### Testing Done

all current tests pass. Simulated tests pass without  warnings from crease_defender_tactic. 
### Resolved Issues
resolves #1805 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/geom/algorithms/contains.cpp [hunk 1]
```diff
@@ -61,13 +62,13 @@ bool contains(const Polygon& container, const Point& contained)
...
+    bool point_is_ray_start      = contained == container.getStart();
     bool point_collinear_with_ray =
-        collinear(contained, container.getStart(), point_in_ray_direction);
-    bool point_is_in_ray_direction =
+        collinear(contained, container.getStart(), point_in_ray_direction, 1e-9, 200);
+    bool point_is_in_ray_direction = almostEqual(
         ((contained - container.getStart()).normalize() - container.toUnitVector())
-            .length() < FIXED_EPSILON;
+            .length(),
+        0, 1e-9, 200);
```
Comment (jonathanlew) [inline src/software/geom/algorithms/contains.cpp]: is it possible to use geom constants or some multiple of one, e.g. `2*FIXED_EPSILON`, so we don't have magic numbers in our code?

**Review by jonathanlew — CHANGES_REQUESTED** [review]
nice finds

File: src/software/geom/algorithms/contains.cpp (hunk 1 above)
Comment (amrm3lm) [reply src/software/geom/algorithms/contains.cpp]: used `1000 * FIXED_EPSILON` and `2 * ULPS_EPSILON_HUNDRED`, which work fine

Comment (amrm3lm) [reply src/software/geom/algorithms/contains.cpp]: used `1000 * FIXED_EPSILON` and `2 * ULPS_EPSILON_HUNDRED`, which work fine

File: src/software/geom/algorithms/contains.cpp [hunk 2]
```diff
@@ -61,13 +62,14 @@ bool contains(const Polygon& container, const Point& contained)
...
     bool point_collinear_with_ray =
-        collinear(contained, container.getStart(), point_in_ray_direction);
-    bool point_is_in_ray_direction =
+        collinear(contained, container.getStart(), point_in_ray_direction,
+                  1000 * FIXED_EPSILON, 2 * ULPS_EPSILON_HUNDRED);
+    bool point_is_in_ray_direction = almostEqual(
         ((contained - container.getStart()).normalize() - container.toUnitVector())
-            .length() < FIXED_EPSILON;
+            .length(),
+        0, 1000 * FIXED_EPSILON, 2 * ULPS_EPSILON_HUNDRED);
```
Comment (akhilveeraghanta) [inline src/software/geom/algorithms/contains.cpp]: could you leave a quick comment why this specific usage of collinear needs to be given the epsilon and ulps manually and why the default won't work?

and are these the smallest values that work?

File: src/software/simulated_tests/defense_play_test.cpp [hunk 3]
```diff
@@ -16,7 +16,7 @@ TEST_F(DefensePlayTest, test_defense_play)
 {
     setBallState(BallState(Point(0.9, 2.85), Vector(0, 0)));
     addFriendlyRobots(TestUtil::createStationaryRobotStatesWithId(
-        {Point(-3, 2.5), Point(-3, 1.5), Point(-3, 0.5), Point(-3, -0.5), Point(-3, -1.5),
+        {Point(-4.5, 0), Point(-3, 1.5), Point(-3, 0.5), Point(-3, -0.5), Point(-3, -1.5),
```
Comment (tarakong01) [inline src/software/simulated_tests/defense_play_test.cpp:19]: Why did you change the first position from (-3, 2.5) to (-4.5, 0) for all of your test cases? Otherwise LGTM :)

Comment (amrm3lm) [reply src/software/simulated_tests/defense_play_test.cpp:19]: The first position refers to the goalkeepers position at the start of the test. `(-3, 2.5)` is outside the `defender_crease`, which is essentially an inflated version of our defense rectangle. For all tests that used the `crease_defender_tactic`, this caused the warning:

`WARNING [crease_defender_tactic.cpp->calculateDesiredState:114] Could not draw a line from the goalie to the ball through the defender crease` 

Changing the goalkeepers position to `(-4.5,0)` means the goalkeeper starts inside the `defender_crease`, so we dont get that warning anymore.

File: src/software/geom/algorithms/contains.cpp (hunk 2 above)
Comment (amrm3lm) [reply src/software/geom/algorithms/contains.cpp]: The differences being compared varied each run and the smallest value that works depended on where I was running the tests from. If I ran from CLion then an fixed_epsilon of `1e-14` was enough, but running from terminal `1e-13` was needed some times. So to be cautious I used `1e-12`. 

For the ulps, most cases the ulps was under 200, but to eliminate all warnings using ulps alone would require something over 5000, so I actually wasn't sure if we should bother finding/using that value as fixed_epsilon was enough to remove warnings, but im open to thoughts?

Comment (akhilveeraghanta) [reply src/software/geom/algorithms/contains.cpp]: Thats reasonable I was just curious, could you leave a quick comment about the `1e-12` and the 200 ULPS, other than that LGTM :+1:

Comment (amrm3lm) [reply src/software/geom/algorithms/contains.cpp]: added comment

**Review by itsarune — DISMISSED** [review]
Looks good!

**Review by tarakong01 — DISMISSED** [review]

File: src/software/geom/algorithms/contains.cpp [hunk 4]
```diff
@@ -61,13 +62,17 @@ bool contains(const Polygon& container, const Point& contained)
 bool contains(const Ray& container, const Point& contained)
 {
     Point point_in_ray_direction = container.getStart() + container.toUnitVector();
+    bool point_is_ray_start      = contained == container.getStart();
 
-    bool point_is_ray_start = contained == container.getStart();
+    // We use broader values for epsilon and ulps because the default values brought
+    // warnings from simulated tests
```
Comment (jonathanlew) [inline src/software/geom/algorithms/contains.cpp]: We shouldn't be designing/tuning contains around how simulated test warnings. I think fixing this in a non-hacky way is pretty tricky, so please revert all the changes in this file and we'll deal with them later.

**Review by jonathanlew — CHANGES_REQUESTED** [review]

File: src/software/geom/algorithms/contains.cpp (hunk 4 above)
Comment (amrm3lm) [reply src/software/geom/algorithms/contains.cpp]: We could make epsilon and ulps default parameters that can be overwritten, like one version of contains does?

Comment (jonathanlew) [reply src/software/geom/algorithms/contains.cpp]: I don't see how you could pass these values all the way down. In any case, I've looked into this and there are actual bugs in the contains function that can affect not just these warnings but also calcbestshot

Comment (amrm3lm) [reply src/software/geom/algorithms/contains.cpp]: fair enough, ill revert the changes and let them be addressed at another time

Comment (amrm3lm) [reply src/software/geom/algorithms/contains.cpp]: removed changes affecting contains.cpp

**Comment by amrm3lm** [comment]
I added a fix in `theta_star_planner.cpp` that will help with the navigator warning we see in the Crease Defender simulated tests

File: src/software/ai/navigator/path_planner/theta_star_path_planner.cpp [hunk 5]
```diff
@@ -182,7 +182,8 @@ std::optional<Path> ThetaStarPathPlanner::findPath(
 
     // if the start and end points are close enough, then return a straightline path
     if ((start - end).length() < CLOSE_TO_END_THRESHOLD ||
-        ((start - end).length() < SIZE_OF_GRID_CELL_IN_METERS))
+        ((abs(start.x() - end.x()) < SIZE_OF_GRID_CELL_IN_METERS) &&
+         abs(start.y() - end.y()) < SIZE_OF_GRID_CELL_IN_METERS))
```
Comment (jonathanlew) [inline src/software/ai/navigator/path_planner/theta_star_path_planner.cpp]: nit for clarity
```suggestion
        ((std::abs(start.x() - end.x()) < SIZE_OF_GRID_CELL_IN_METERS) &&
         std::abs(start.y() - end.y()) < SIZE_OF_GRID_CELL_IN_METERS))
```

**Review by jonathanlew — COMMENTED** [review]
Looks good, just one more thing

**Review by jonathanlew — APPROVED** [review]
LGTM! :+1:

**Review by jared-paul — APPROVED** [review]
LGTM! 🎉 💯
